sotoki/sotoki.py

```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
+
+        # check if autogenerated identicon from gravatar
+        # for more details see https://en.gravatar.com/site/implement/images/
+        if query.get("d", query.get("default", "")) == "identicon" and query.get(
+            "f", query.get("forcedefault", "")
+        ) in ["y", "1"]:
+            return None, True
+
+        # check for mystery-person type identicon
+        if query.get("d", query.get("default", "")) == "mp":
+            return (
+                download_and_return_name(
+                    url, "gravatar_mystery", convert_png, resize, ext
+                ),
+                False,
+            )
+
+        # check for blank identicon
+        if query.get("d", query.get("default", "")) == "blank":
+            return (
+                download_and_return_name(
+                    url, "gravatar_blank", convert_png, resize, ext
+                ),
+                False,
+            )
+
+        # update query with size = 128 (its the size in which we download identicons)
+        query.update({"s": "128"})
+        url_parts[4] = urllib.parse.urlencode(query)
+        url_with_size = urllib.parse.urlunparse(url_parts)
+
+        # check if potentially a duplicate default image at size = 128
+        headers = get_response_headers(url_with_size)
+        if headers.get("content-length") == "4268":
+            return (
+                download_and_return_name(
+                    url, "gravatar_default1", convert_png, resize, ext
+                ),
+                False,
+            )
+        elif headers.get("content-length") == "3505":
+            return (
+                download_and_return_name(
+                    url, "gravatar_default2", convert_png, resize, ext
+                ),
+                False,
+            )
+
+    if "googleusercontent" in parsed_url.netloc:
+        headers = get_response_headers(url)
+
+        # custom images have etag whereas autogenerated ones do not (mostly)
+        if not headers.get("etag", None):
+            content_length = headers.get("content-length")
+            image_name_prefix = f"googleusercontent_{content_length}"
+            return (
+                download_and_return_name(
+                    url, image_name_prefix, convert_png, resize, ext
+                ),
+                False,
+            )
+
+    if "blend-exchange" in parsed_url.netloc:
+        # a specific image is duplicated, whatever the query part
+        if parsed_url.path == "/embedImage.png":
+            return (
+                download_and_return_name(
+                    url, "blend_exchange", convert_png, resize, ext
+                ),
+                False,
+            )
+
+    return None, None
+
+
+def process_potential_duplicate(url, fullpath, convert_png, resize):
+    """ write a redirection for duplicate files or skip download if we can generate similar identicon
+        reyurns true if redirection was written or download can be skipped """
```

Comment:
```py
def handle_user_identicon(url, fullpath, convert_png, resize):
    """ Whether identicon is duplicate of existing (and process accordingly)
         
         Download identicon file if not (yet) a duplicate
         Write a redirection entry if a duplicate """
```

Reply:
Named it handle_duplicate_images() as it's not just for identicons, though most of the effect is on identicons

Reply:
Got it